lib/gitsh/completion_escaper.rb

```diff
@@ -0,0 +1,46 @@
+require 'gitsh/parser'
+
+module Gitsh
+  class CompletionEscaper
+    ESCAPABLES = {
+      nil => Gitsh::Parser::UNQUOTED_STRING_ESCAPABLES,
+      '"' => Gitsh::Parser::SOFT_STRING_ESCAPABLES,
+      "'" => Gitsh::Parser::HARD_STRING_ESCAPABLES,
+    }.freeze
+
+    def initialize(completer, options)
+      @completer = completer
+      @line_editor = options[:line_editor]
+    end
+
+    def call(input)
+      completer.call(unescape(input)).map { |option| escape(option) }
+    end
+
+    private
+
+    attr_reader :completer, :line_editor
+
+    def escape(option)
+      escaped = option.gsub(/([#{escapables}])(?!$)/) { |char| "\\#{char}" }
+
+      if completing_quoted_argument?
```

Comment:
This is a little unpleasant, and is refactored out in #260. I didn't include that refactoring in this PR because it touched a few different classes, and this PR is already pretty large.